### DISCUSSION TITLE
Missing fabm0d executable

### DISCUSSION THREAD

**Discussion by platipodium** [fabm-model/fabm#49] [Q&A] [answered]
Dear Jorn and FABM community,

I followed the instructions on the FABM wiki for building the 0d host at https://github.com/fabm-model/fabm/wiki/0D-driver

The build instructions are:

```
mkdir build
cd build
cmake <FABMDIR>/src/drivers/0d -DGOTM_BASE=<GOTMDIR>
cmake --build . --target install
```

and should result in a full installation in `~/local/fabm/0d/`.  For me and others, the  `include`, `lib`, and  `cmake` installed directories are present and complete.  But: the  with `bin/fabm0d` executable is missing.  How do I obtain this executable?

**Comment by platipodium** [comment]
Tagging @kaiwirtz3

**Comment by jornbr** [comment]
Hi Carsten,

I'm not immediately sure what's going on there. We verify that the fabm0d executable is created as part of our automated testing on Linux, and there [it currently succeeds](https://github.com/fabm-model/fabm/actions/runs/7487884022/job/20381196846#step:5:255). The simplest explanation would be if you ran cmake differently, with `cmake <FABMDIR> -DFABM_HOST=0d`, but the instructions you cite above should be ok... What's the final output from your `cmake --build . --target install` command? it should mention which files it is installing, and where.

Cheers,

Jorn

**Comment by kaiwirtz3** [comment]
hi jorn,

the command 
% cmake --build . --target install

results in the output below, thus lacks any "bin" directory or "fabm0d" executable.
any help is much appreciated
best, kai
 
[  7%] Built target fabm_base
[  9%] Built target yaml
[ 29%] Built target fabm_models_aed
[ 32%] Built target fabm_models_iow
[ 33%] Built target fabm_models_klimacampus
[ 34%] Built target fabm_models_metu
[ 35%] Built target fabm_models_msi
[ 40%] Built target fabm_models_au
[ 44%] Built target fabm_models_bb
[ 54%] Built target fabm_models_examples
[ 58%] Built target fabm_models_gotm
[ 77%] Built target fabm_models_hzg
[ 81%] Built target fabm_models_jrc
[ 84%] Built target fabm_models_pml
[ 90%] Built target fabm_models_niva
[ 93%] Built target fabm_models_akvaplan
[100%] Built target fabm
Install the project...
-- Install configuration: "Release"
-- Up-to-date: /Users/wirtz/local/fabm/0d/lib/libfabm.a
-- Up-to-date: /Users/wirtz/local/fabm/0d/include
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_types.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_coupling.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_config_types.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_yaml.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_properties.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_standard_variables.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_version.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_expressions.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_config.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_driver.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_particle.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_library.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_parameters.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_schedule.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_builtin_models.mod
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm.h
-- Up-to-date: /Users/wirtz/local/fabm/0d/include/fabm_driver.h
-- Up-to-date: /Users/wirtz/local/fabm/0d/cmake/fabmConfig.cmake
-- Up-to-date: /Users/wirtz/local/fabm/0d/cmake/fabmConfig-release.cmake

**Comment by jornbr** [comment]
Hi Kai,

Can you try `cat build/CMakeCache.txt | grep -E SOURCE_DIR` and post its output here? It should include a line that starts with `fabm0d_SOURCE_DIR:STATIC=`. If that's not there, something must have gone wrong with the very first cmake command. In that case, I'd suggest deleting your `build` directory and rerunning the installation instructions above from scratch. If that does not result in `-- Installing: ~/local/fabm/0d/bin/fabm0d`, please post the full output of all commands here.

Cheers,

Jorn

**Comment by platipodium** [comment]
Hallo Jorn,

```
$ cat CMakeCache.txt | grep -E SOURCE_DIR
fabm_SOURCE_DIR:STATIC=/Users/Lemmen/devel/fabm/fabm-official
yaml_SOURCE_DIR:STATIC=/Users/Lemmen/devel/fabm/fabm-official/src/yaml
```

Also fehlt die gewünschte Zeile `fabm0d_SOURCE_DIR:STATIC=. `.

Der Output des ersten CMAKE ist:

```
$  cmake .. -DFABM_HOST=0d
-- The Fortran compiler identification is GNU 13.2.0
-- Checking whether Fortran compiler has -isysroot
-- Checking whether Fortran compiler has -isysroot - yes
-- Checking whether Fortran compiler supports OSX deployment target flag
-- Checking whether Fortran compiler supports OSX deployment target flag - yes
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Check for working Fortran compiler: /opt/local/bin/gfortran - skipped
-- FABM host: 0d
-- Installation prefix: /Users/Lemmen/local/fabm/0d
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/Lemmen/devel/fabm/fabm-official/build
 ```

**Comment by jornbr** [comment, marked as answer]
Hi Carsten,

Both those indicate that you ran cmake with `<FABMDIR>` as source dir, rather than `<FABMDIR>/src/drivers/0d`. If you use `cmake ../src/drivers/0d -DGOTM_BASE=<GOTMDIR>`, you're likely ok.

**Comment by platipodium** [comment]
You're right, that's what we should have done.  Only that GOTM compilation fails (with up-to-date sources) when doing that.

```
cmake ../src/drivers/0d -DGOTM_BASE=~/devel/gotm && cmake --build . --target install
```

```
...
[ 98%] Building Fortran object CMakeFiles/fabm0d.dir/fabm0d.F90.o
/Users/Lemmen/devel/fabm/fabm-official/src/drivers/0d/fabm0d.F90:17:8:

   17 |    use eqstate,only:rho_feistel
      |        1
Fatal Error: Cannot open module file 'eqstate.mod' for reading at (1): No such file or directory
```

but I guess that is another discussion.  I guess it should be made clearer in the documentation of the FABM0d host that it is insufficient to do `cmake $FABMDIR -DFABM_HOST=0d`, but instead that it is required to use `cmake $FABMDIR/src/drivers/0d`  ...

**Comment by kaiwirtz3** [comment]
hi jorn,

carsten and me (on macbooks with Ports installations) share the same problems.

cmake ../code/src/drivers/0d -DGOTM_BASE=/Users/wirtz/mossco/code/external/gotm/code

CMake Error at CMakeLists.txt:52 (add_executable):
  Cannot find source file:

    /Users/wirtz/mossco/code/external/gotm/code/src/airsea/short_wave_radiation.F90
    
in the most recent gotm pull I have "shortwave_radiation.F90" (without first "_").

****
after creating a softlink "short_wave_radiation.F90" that points to "shortwave_radiation.F90", the first cmake gets through.
hover, the second cmake ("cmake --build . --target install") runs into

[ 94%] Building Fortran object CMakeFiles/fabm0d.dir/fabm0d.F90.o
/Users/wirtz/mossco/code/external/fabm/code/src/drivers/0d/fabm0d.F90:583:77:

  583 |          call register_input_0d(path,column,pdata,variable_name,scale_factor=scale_factor)
      |                                                                             1
Error: Keyword argument requires explicit interface for procedure 'register_input_0d' at (1)
gmake[2]: *** [CMakeFiles/fabm0d.dir/build.make:88: CMakeFiles/fabm0d.dir/fabm0d.F90.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:426: CMakeFiles/fabm0d.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2

**Comment by jornbr** [comment]
Glad there's progress! Both aspects (the cmake syntax and which GOTM version to use) are mentioned at [the wiki page you referred to above](https://github.com/fabm-model/fabm/wiki/0D-driver), but if you have ideas how this could be made clearer, they would be very welcome. :-) Also, if you found incorrect/outdated instructions elsewhere, let me know.

NB to pull in the required stable GOTM, you could do `git clone --recursive -b v6.0 https://github.com/gotm-model/code.git gotm6` (or download the stable release including submodules as mentioned on the wiki)

**Comment by platipodium** [comment]
Dear @jornbr, thanks a lot for helping us to make it work!  Changing the gotm from development to v6.0 stable (`git clone --recursive -b v6.0 https://github.com/gotm-model/code.git`) resolved the above compilation issues.  

I now have a working `~/local/fabm/0d/bin/fabm0d` binary!

My THM for fabm0d
- make sure give to cmake the source path to the `drivers/0d`
- use GOTM version 6.0

**Reply by platipodium** [reply]
Just a side remark @kaiwirtz3 : if you use the outdated FABM0 version with the MOSSCO coupling framework, make sure to use GOTM v5.3.
